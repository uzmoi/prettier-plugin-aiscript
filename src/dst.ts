export interface Loc {
	start: number;
	end: number;
}

export interface Comment extends NodeBase {
	type: "Comment";
	value: string;
	nodeDescription?: string;
	placement?: "ownLine" | "endOfLine" | "remaining";
	leading?: boolean;
	trailing?: boolean;
	printed?: boolean;
}

export interface NodeBase<T extends string = string> {
	type: T;
	loc: Loc;
	comments?: Comment[];
}

interface FnBase extends NodeBase {
	params: FnParameter[];
	returnTy: Ty | null;
	body: Block;
}

export interface FnParameter extends NodeBase {
	type: "FnParameter";
	name: Identifier;
	ty: Ty | null;
	optional: boolean;
	default: Expression | null;
}

export type Node = Script | TopLevel | Statement | Expression | Ty | Comment;

export interface Script extends NodeBase {
	type: "Script";
	body: (TopLevel | Statement)[];
}

export type TopLevel = Namespace | Meta;

export interface Namespace extends NodeBase {
	type: "Namespace";
	name: Identifier;
	body: NamespaceItem[];
}

export type NamespaceItem = Namespace | VariableDefinition | FnDefinition;

export interface Meta extends NodeBase {
	type: "Meta";
	name: Identifier | null;
	value: Expression;
}

export type Statement =
	| Assignment
	| VariableDefinition
	| FnDefinition
	| Return
	| Each
	| For
	| Loop
	| While
	| Break
	| Continue
	| Out
	| ExpressionStatement
	| Block;

export interface Assignment extends NodeBase {
	type: "Assignment";
	operator: "=" | "+=" | "-=";
	dest: Expression;
	value: Expression;
}

export interface VariableDefinition extends NodeBase {
	type: "VariableDefinition";
	mutable: boolean;
	name: Identifier;
	ty: Ty | null;
	init: Expression;
}

export interface FnDefinition extends FnBase {
	type: "FnDefinition";
	name: Identifier;
}

export interface Return extends NodeBase {
	type: "Return";
	body: Expression;
}

export interface Each extends NodeBase {
	type: "Each";
	definition: VariableDefinition;
	source: Expression;
	body: Statement;
}

export interface For extends NodeBase {
	type: "For";
	enumerator:
		| { type: "Times"; times: Expression }
		| {
				type: "Range";
				definition: VariableDefinition;
				from: Expression | null;
				to: Expression;
		  };
	body: Statement;
}

export interface Loop extends NodeBase {
	type: "Loop";
	body: Block;
}

export interface While extends NodeBase {
	type: "While";
	do: boolean;
	condition: Expression;
	body: Statement;
}

export interface Break extends NodeBase {
	type: "Break";
}

export interface Continue extends NodeBase {
	type: "Continue";
}

export interface Out extends NodeBase {
	type: "Out";
	body: Expression;
}

export interface ExpressionStatement extends NodeBase {
	type: "ExpressionStatement";
	expression: Expression;
}

export interface Block extends NodeBase {
	type: "Block";
	body: Statement[];
}

export type Expression =
	| Identifier
	| Literal
	| Template
	| UnaryOperator
	| BinaryOperator
	| Fn
	| Call
	| Index
	| Prop
	| If
	| Match
	| EvalBlock;

export interface Identifier extends NodeBase {
	type: "Identifier";
	name: string;
}

export type Literal =
	| NullLiteral
	| BoolLiteral
	| NumberLiteral
	| StringLiteral
	| ArrayLiteral
	| ObjectLiteral;

export interface NullLiteral extends NodeBase {
	type: "NullLiteral";
}

export interface BoolLiteral extends NodeBase {
	type: "BoolLiteral";
	value: boolean;
}

export interface NumberLiteral extends NodeBase {
	type: "NumberLiteral";
	value: number;
}

export interface StringLiteral extends NodeBase {
	type: "StringLiteral";
	value: string;
}

export interface Template extends NodeBase {
	type: "Template";
	parts: (TemplatePart | Expression)[];
}

export interface TemplatePart extends NodeBase {
	type: "TemplatePart";
	content: string;
}

export interface ArrayLiteral extends NodeBase {
	type: "ArrayLiteral";
	elements: Expression[];
}

export interface ObjectLiteral extends NodeBase {
	type: "ObjectLiteral";
	properties: ObjectProperty[];
}

export interface ObjectProperty extends NodeBase {
	type: "ObjectProperty";
	key: Identifier;
	value: Expression;
}

export interface UnaryOperator extends NodeBase {
	type: "UnaryOperator";
	operator: "!" | "+" | "-" | "exists";
	body: Expression;
}

export interface BinaryOperator extends NodeBase {
	type: "BinaryOperator";
	operator:
		| ("&&" | "||")
		| ("==" | "!=")
		| (">" | ">=" | "<" | "<=")
		| ("+" | "-" | "*" | "/" | "%" | "^");
	rhs: Expression;
	lhs: Expression;
}

export interface Fn extends FnBase {
	type: "Fn";
}

export interface Call extends NodeBase {
	type: "Call";
	callee: Expression;
	args: Expression[];
}

export interface Index extends NodeBase {
	type: "Index";
	target: Expression;
	index: Expression;
}

export interface Prop extends NodeBase {
	type: "Prop";
	target: Expression;
	name: Identifier;
}

export interface If extends NodeBase {
	type: "If";
	condition: Expression;
	then: Statement;
	elseif: ElseIf[];
	else: Statement | null;
}

export interface ElseIf extends NodeBase {
	type: "ElseIf";
	condition: Expression;
	then: Statement;
}

export interface Match extends NodeBase {
	type: "Match";
	value: Expression;
	cases: MatchCase[];
}

export interface MatchCase extends NodeBase {
	type: "MatchCase";
	pattern: Expression | null;
	body: Statement;
}

export interface EvalBlock extends NodeBase {
	type: "EvalBlock";
	body: Block;
}

export type Ty = TypeReference | FnType;

export interface TypeReference extends NodeBase {
	type: "TypeReference";
	name: Identifier;
	argument: Ty | null;
}

export interface FnType extends NodeBase {
	type: "FnType";
	params: Ty[];
	return: Ty;
}
