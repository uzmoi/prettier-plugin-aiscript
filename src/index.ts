import prettier from "prettier";
import { printAiScript } from "./printer";
import type { Comment, Node } from "./node";
import { parse } from "./parse";

const parser: prettier.Parser<Node> = {
	parse(text, _options) {
		return parse(text);
	},
	astFormat: "aiscript",
	locStart(node) {
		return node.loc?.start ?? 0;
	},
	locEnd(node) {
		return node.loc?.end ?? 0;
	},
};

const printer: prettier.Printer<Node> = {
	print(path, options, print, _args) {
		return printAiScript(path, options, print);
	},
	hasPrettierIgnore(path) {
		const { node } = path;

		if (!node.comments) {
			return false;
		}

		return node.comments.some(comment =>
			/\/[/*]\s*prettier-ignore/.test(comment.value),
		);
	},
	canAttachComment(node) {
		return (node as { type: string }).type !== "comment";
	},
	printComment(commentPath, _options) {
		const node = commentPath.node as unknown as Comment;
		return node.type === "comment" ? node.value : "";
	},
	getCommentChildNodes(node, _options) {
		switch (node.type) {
			case "if":
				return [
					node.cond,
					node.then,
					...node.elseif.flatMap(({ cond, then }) => [cond, then]),
					...(node.else ? [node.else] : []),
				];
			case "match":
				return [
					node.about,
					...node.qs.flatMap(q => [q.q, q.a]),
					...(node.default ? [node.default] : []),
				];
			case "obj":
				return [...node.value.values()];
		}
	},
};

const plugin: prettier.Plugin = {
	languages: [
		{
			name: "AiScript",
			parsers: ["aiscript"],
			extensions: ["is"],
		},
	],
	parsers: { aiscript: parser },
	printers: { aiscript: printer },
};

export default plugin;
