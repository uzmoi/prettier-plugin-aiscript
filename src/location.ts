import type * as dst from "./dst";

export const locStart = (node: dst.NodeBase): number => {
	return node.loc.start;
};

export const locEnd = (node: dst.NodeBase): number => {
	return node.loc.end;
};
