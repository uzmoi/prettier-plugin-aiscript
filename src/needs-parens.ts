import type { ParserOptions } from "prettier";
import type { Ast } from "@syuilo/aiscript";
import type { Node } from "./node";
import type { AstPath } from "./types";
import { isSugarCall, type SugarCall } from "./sugar";

// https://github.com/aiscript-dev/aiscript/blob/9e618049b5753b26d7527ee736dff10d65289b18/src/parser/plugins/infix-to-fncall.ts#L92-L133
const opPrecedenceTable = {
	"Core:mul": 7,
	"Core:pow": 7,
	"Core:div": 7,
	"Core:mod": 7,
	"Core:add": 6,
	"Core:sub": 6,
	"Core:eq": 4,
	"Core:neq": 4,
	"Core:lt": 4,
	"Core:gt": 4,
	"Core:lteq": 4,
	"Core:gteq": 4,
	and: 3,
	or: 3,
} as const;

type BinaryOperatorNode = Ast.And | Ast.Or | SugarCall;

const isBinaryOperator = (
	node: Node,
	options: ParserOptions,
): node is BinaryOperatorNode =>
	node.type === "and" || node.type === "or" || isSugarCall(node, options);

const getPrecedence = (node: BinaryOperatorNode): number =>
	opPrecedenceTable[node.type === "call" ? node.target.name : node.type];

const isRhs = (node: Node, parent: BinaryOperatorNode): boolean =>
	(parent.type === "call" ? parent.args[1] : parent.right) === node;

const isTrailing = (parent: Node, key: string | null): boolean =>
	parent.type === "prop" ||
	parent.type === "index" ||
	(parent.type === "call" && key === "target");

export const needsParens = (path: AstPath, options: ParserOptions): boolean => {
	const { node, parent, key } = path;

	if (parent == null) return false;

	switch (node.type) {
		case "not":
		case "fn":
		case "exists":
			return isTrailing(parent, key) || isBinaryOperator(parent, options);
	}

	// HACK: これがないと型が正常に絞り込まれない。
	// これがあると何故かisBinaryOperatorでちゃんと絞り込みが効く。
	// TypeScript何もわからん
	if (import.meta.vitest) {
		isSugarCall(node, options) ? 0 : 1;
		isSugarCall(parent, options) ? 0 : 1;
	}

	if (isBinaryOperator(node, options)) {
		if (import.meta.vitest) {
			const { expectTypeOf } = import.meta.vitest;
			expectTypeOf(node.type).extract<"call">().not.toBeNever();
		}

		if (isTrailing(parent, key)) return true;

		if (isBinaryOperator(parent, options)) {
			if (import.meta.vitest) {
				const { expectTypeOf } = import.meta.vitest;
				expectTypeOf(parent.type).extract<"call">().not.toBeNever();
			}

			const precedence = getPrecedence(node);
			const parentPrecedence = getPrecedence(parent);

			if (
				parentPrecedence > precedence ||
				(isRhs(node, parent) && parentPrecedence === precedence)
			) {
				return true;
			}
		}
	}

	return false;
};
