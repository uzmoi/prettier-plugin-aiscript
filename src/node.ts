import type { Ast } from "@syuilo/aiscript";

export type Comment = {
	type: "comment";
	loc: Ast.Loc;
	value: string;
};

export type Root = {
	type: "root";
	body: Ast.Node[];
	loc?: Ast.Loc;
	comments: Comment[];
};

export type Node =
	| (Ast.Node & { comments?: Comment[]; sugar?: boolean })
	| Root;
