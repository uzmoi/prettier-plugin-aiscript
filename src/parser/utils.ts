import type { Cst } from "@syuilo/aiscript";
import { visitNode } from "@syuilo/aiscript/parser/visit.js";

export const parserPlugin =
	(f: (node: Cst.Node) => Cst.Node) => (nodes: Cst.Node[]) =>
		nodes.map(node => visitNode(node, f));
