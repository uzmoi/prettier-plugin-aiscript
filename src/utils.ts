import type { ParserOptions } from "prettier";
import type { Node } from "./node";

// export const getNodeSourceCode = (node: Node, options: ParserOptions<Node>) => {
// 	const { originalText, locStart, locEnd } = options;
// 	return originalText.slice(locStart(node), locEnd(node));
// };

export const startsWith = (
	string: string,
	node: Node,
	options: ParserOptions<Node>,
) => {
	const { originalText, locStart } = options;
	return originalText.startsWith(string, locStart(node));
};
