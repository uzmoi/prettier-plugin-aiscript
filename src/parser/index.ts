import { AISCRIPT_VERSION, type Ast, Parser } from "@syuilo/aiscript";
import { cmp, coerce, satisfies } from "semver";
import type { Root } from "../node";
import {
	correctLocation,
	parseCommentsByPreprocessDiff,
	parseCommentsByStringLocations,
} from "./parse-comments";
import { transformChainPlugin } from "./transform-chain";
import { parserPlugin } from "./utils";

export const emptyNsPlugin = () =>
	parserPlugin(node => {
		if (node.type === "ns") {
			node.members ??= [];
		}
		return node;
	});

const parse_0_19_0 = (source: string): Root => {
	const parser = new Parser();

	// @ts-expect-error
	parser.plugins.validate.unshift(emptyNsPlugin());

	parser.addPlugin("transform", transformChainPlugin());
	parser.addPlugin(
		"transform",
		parserPlugin(node => {
			if (node.loc == null) return node;
			const { start, end } = node.loc;
			// AiScriptのパーサーがendに最後の文字のindexを入れるので+1する。
			return { ...node, loc: { start, end: end + 1 } };
		}),
	);

	const stringLocations: Ast.Loc[] = [];
	parser.addPlugin(
		"transform",
		// collect string locations
		parserPlugin(node => {
			const isStringNode = node.type === "str" || node.type === "tmpl";
			if (isStringNode && node.loc) {
				stringLocations.push(node.loc);
			}
			return node;
		}),
	);

	const body = parser.parse(source);

	const comments = parseCommentsByStringLocations(source, stringLocations);

	return { type: "root", body, comments };
};

const parse_before_0_19_0 = (source: string): Root => {
	const comments = parseCommentsByPreprocessDiff(source);

	const parser = new Parser();

	// @ts-expect-error
	parser.plugins.validate.unshift(emptyNsPlugin());

	parser.addPlugin("transform", transformChainPlugin());
	parser.addPlugin(
		"transform",
		parserPlugin(node => correctLocation(node, comments)),
	);

	const body = parser.parse(source);

	return { type: "root", body, comments };
};

export const parse = (text: string): Root => {
	const version = coerce(AISCRIPT_VERSION)!;

	if (satisfies(version, "0.19")) {
		return parse_0_19_0(text);
	}

	if (cmp(version, "<", "0.19.0")) {
		return parse_before_0_19_0(text);
	}

	throw new Error(`Unsupported AiScript version: ${AISCRIPT_VERSION}`);
};
