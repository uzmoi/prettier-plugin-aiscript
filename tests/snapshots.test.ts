import { Parser } from "@syuilo/aiscript";
import { describe, test } from "vitest";
import { emptyNsPlugin } from "../src/parser";
import { format, parserPlugin } from "./utils";

const cases = import.meta.glob<string>("./src/*.ais", {
	eager: true,
	query: "raw",
	import: "default",
});

const parser = new Parser();

// @ts-expect-error
parser.plugins.validate.unshift(emptyNsPlugin());

parser.addPlugin(
	"transform",
	// @ts-expect-error
	// biome-ignore lint: in test.
	parserPlugin(node => (delete node.loc, delete node.operatorLoc, node)),
);

describe.concurrent.each(
	Object.entries(cases).map(([path, source]) => {
		const name = path.replace(/^.*\/([^/]+)\.ais$/, "$1");
		return [name, source] as const;
	}),
)("%s", (_, source) => {
	test("snapshot", async ({ expect }) => {
		expect(await format(source)).toMatchSnapshot();
	});

	test("ast no change", async ({ expect }) => {
		const formatted = await format(source);
		expect(parser.parse(formatted)).toStrictEqual(parser.parse(source));
	});
});
