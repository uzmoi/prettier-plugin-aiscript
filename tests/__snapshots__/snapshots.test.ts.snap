// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`call > snapshot 1`] = `
"f()

f(0)

f(0, 1)

f(@() {
  //
})

f(0, 1, @() {
  //
})

f(
  super_ultra_hyper_mega_miracle_ultimate_extremely_very_long_argument_expression,
  @() {
    //
  }
)

f({ en: ["foo", "bar", "baz", "qux"], ja: ["hoge", "fuga", "piyo"] }, @(a) {
  a
})

f(@(array) {
  each (let element, array) {
    <: \`type = {Core:type(element)}, value = {element}\`
  }
})

f(
  {
    hoge: 1,
  },
  last_argument
)
"
`;

exports[`literals > snapshot 1`] = `
"null

true
false

42
6.28

"Hello AiScript!"
'"'

\`Hello {name}!\`
\`\\\`\\{\`
"
`;

exports[`toplevel > snapshot 1`] = `
"### {}

### {
  hoge: 0,
}

###  hoge{}

:: Empty {}

:: A {
  let hoge = 0

  :: B {
    let fuga = 2
    let piyo = 3
  }

  :: C { let foo = -1 }
}
"
`;
